scripts/speech_recognition/estimate_duration_bins_2d.py

```diff
@@ -0,0 +1,319 @@
+        "-q", "--quiet", type=bool, default=False, help="When specified, only print the estimated duration bins."
+    )
+    parser.add_argument(
+        "-f",
+        "--prompt-format",
+        type=str,
+        help="When specified, we'll use a prompt formatter in addition to the tokenizer for the purpose of estimating token count bins. "
+        "This is useful for accurate 2D bucket estimation with models such as EncDecMultiTaskModel (Canary-1B), "
+        "or any model where the label sequence consists of a user prompt and a model's response.",
+    )
+    parser.add_argument(
+        "-p",
+        "--prompt",
+        type=str,
+        help="Prompt slots provided as a Python list of dicts. It is used together with --prompt-format option."
+        "For example, with Canary-1B you may use: [{'role':'user','slots':{'source_lang':'en','target_lang':'en','task':'asr','pnc':'yes'}]",
+    )
+    return parser.parse_args()
+
+
+def estimate_duration_buckets(
+    cuts: Iterable[Cut],
+    num_buckets: int,
+    num_subbuckets: int,
+    max_tps: float,
+    max_duration: float,
+    quiet: bool,
+) -> list[tuple[float, float]]:
+    assert num_buckets > 1
+
+    constraint = FixedBucketBatchSizeConstraint2D([(0.0, 0.0)], [0])
+
+    sizes = []
+    num_tokens = []
+    for c in cuts:
+        dur, toks = constraint.measure_length(c)
+        sizes.append(dur)
+        num_tokens.append(toks)
+    sizes = np.array(sizes, dtype=np.float32)
+    num_tokens = np.array(num_tokens, dtype=np.int32)
+    joint = np.rec.fromarrays([sizes, num_tokens])
+    joint.sort()
+    sizes = joint.f0
+    num_tokens = joint.f1
+
+    size_per_bucket = sizes.sum() / num_buckets
+
+    if not quiet:
+        print("Duration distribution:")
+        print(pd.Series(sizes).describe(percentiles=[0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99]))
+    if math.isinf(max_duration):
+        max_duration = sizes[-1]
+
+    tps = num_tokens / sizes
+    if not quiet:
+        print("Token per second distribution:")
+        print(pd.Series(tps).describe(percentiles=[0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99]))
+    if math.isinf(max_tps):
+        max_tps = tps.max()
+    del tps
+
+    bins = []
+    bin_indexes = [0]
+    tot = 0.0
+    for binidx, size in enumerate(sizes):
+        if tot > size_per_bucket:
+            num_tokens_bucket = num_tokens[bin_indexes[-1] : binidx]
+            num_tokens_bucket.sort()
+            tokens_per_subbucket = num_tokens_bucket.sum() / num_subbuckets
+            tot_toks = 0
+            for num_toks in num_tokens_bucket:
+                if tot_toks > tokens_per_subbucket:
+                    bins.append((size, num_toks))
+                    tot_toks = 0
+                tot_toks += num_toks
+            bins.append((size, math.ceil(size * max_tps)))
+            bin_indexes.append(binidx)
+            tot = 0.0
+        tot += size
+
+    num_tokens_bucket = num_tokens[bin_indexes[-1] : binidx]
```

Comment:
Consider somehow merging this into the loop above. Not sure what that would take. Up to you.

Reply:
Refactored using a closure